### ISSUE TITLE
Strange behaviour with multiple instances of Demo module

### ISSUE THREAD

**Issue by landgrvi** [Paul-Dempsey/svg_theme#3]
I've built the Demo module, and it works as expected when there's a single instance of it. When I add more than one instance, though, things get weird.

For example:
Starting with a blank rack, I add the first Demo. It has the light theme and the menu has 'Light' checked. I change it to dark. It now displays the dark theme, and the menu has 'Dark' checked. So far so good!
I add a second Demo module, and it appears in Rack displaying the dark theme. Its menu, though, has 'Light' checked. 

![image](https://github.com/user-attachments/assets/6fe7db43-3ecf-4758-ba54-3240260e1b48)

When I click 'Light' in its menu, it displays the light theme, as you'd expect.

![image](https://github.com/user-attachments/assets/3b51667a-6111-421e-ba65-5d9b6bd24a9d)

But now, if I close and reopen Rack, both modules display the dark theme, even though 'Light' is checked in the second module's menu. Similar stuff happens as I add more Demo modules but the behaviour gets hard to track. Basically it seems that the menus are independent, but the display theme is somehow not; and closing and reopening Rack sets them all (or some subgroup of them) to the same display theme even when that doesn't match their menus.

![image](https://github.com/user-attachments/assets/a824555d-4eef-423c-992a-775a0799de76)

And after reopening:

![image](https://github.com/user-attachments/assets/3d9787c8-dbbe-4069-9cae-aab46039a99b)

I'm using Rack 2.5.2, built from source, on ubuntu studio 24.04, if that makes a difference.

**Comment by landgrvi** [comment]
OK, the plot thickens. I did a little more experimenting and things got weirder.
I start with a blank Rack and add a Demo module, which defaults to Light.
Then I add a second Demo, which also defaults to Light.
I change either Demo to Dark. I move the other Demo, without touching its menu, and it also draws as Dark!
Like this:

https://github.com/user-attachments/assets/893aefc7-eb39-4f28-ad56-26cdafe14c5c


I'm really confused.

**Comment by landgrvi** [comment]
As you posited over on the VCVRack community forum, it's definitely the SVG cache, which is indexed by filename. I tried overloading Svg::load in the Rack code to allow an additional suffix:
```
std::shared_ptr<Svg> Svg::load(const std::string& filename, const std::string& suffix) {
	const auto& pair = svgCache.find(filename+suffix);
	if (pair != svgCache.end())
		return pair->second;

	// Load svg
	std::shared_ptr<Svg> svg;
	try {
		svg = std::make_shared<Svg>();
		svg->loadFile(filename);
	}
	catch (Exception& e) {
		WARN("%s", e.what());
		svg = NULL;
	}
	svgCache[filename+suffix] = svg;
	return svg;
}
```
and similarly with createPanel:
```
template <class TPanel = app::ThemedSvgPanel>
TPanel* createPanel(std::string lightSvgPath, std::string darkSvgPath, std::string suffix) {
	TPanel* panel = new TPanel;
	panel->setBackground(window::Svg::load(lightSvgPath, suffix), window::Svg::load(darkSvgPath, suffix));
	return panel;
}
```
And was able to use those to make the problem go away by generating a random string as a suffix. An example from my code:
```
struct VGLabsSlider : SvgSlider, svg_theme::IApplyTheme {
	VGLabsSlider() {
		setBackgroundSvg(Svg::load(asset::plugin(pluginInstance, "res/VGLabsSlider.svg"), generateRandomString(8)));
		setHandleSvg(Svg::load(asset::plugin(pluginInstance, "res/VGLabsSliderHandle.svg"), generateRandomString(8)));
		setHandlePosCentered(math::Vec(box.size.x / 2, box.size.y - 5), math::Vec(box.size.x / 2, 5));
	}
```

I wouldn't call this a _solution_ as it did require rebuilding Rack, but it does confirm where the problem is.

**Comment by Paul-Dempsey** [comment]
That was going above and beyond - great job!

Unfortunately a fix compatible with an unmodified Rack is not going to be a simple or easy one. The scheme for wiring up a module will need to change quite a bit. The actual theme application code is fine.

I am thinking a helper that will create a foo-theme.svg in a suitable folder by copying the master svg. Modules then use this file as their module/widget svg to take advantage of Rack SVG caching. Theme switching can no longer modify in-place because that changes the cached svg -- it needs to replace the widget's svg. With luck we can optimize the theme application code to recognize when the theme has already been applied, to take further advantage of the cache.

**Comment by Doom2fan** [comment]
I think an alternative solution that doesn't require a helper could be to create a custom version of Svg::load that takes in the SVG path along with the theme, and then caches the SVG based on the SVG + theme combination. It would mean adding your own caching, but I think that'd be fine for this use case.

**Comment by landgrvi** [comment]
Well, but applying the theme changes the SVG in memory. Which is good, that's the whole point, but means that we wouldn't want to cache based on SVG + theme. The problem happens when different theme holders stomp on each other, so they're what has to be distinguished in the cache. Does the theme holder have a unique id? Does a widget? If not, I think the theme holder could generate one.

I made some attempts at trying to make a custom Svg::load without modifying any of the Rack files, but ran into scoping problems or various flavours thereof (and that's when I gave up and just started recompiling Rack, heh). Something about that static svgCache in Svg.cpp is throwing me for a loop. It's highly likely that I just don't know what I'm doing; my c++ isn't very sophisticated.

**Comment by Doom2fan** [comment]
> Well, but applying the theme changes the SVG in memory. Which is good, that's the whole point, but means that we wouldn't want to cache based on SVG + theme. The problem happens when different theme holders stomp on each other, so they're what has to be distinguished in the cache. Does the theme holder have a unique id? Does a widget? If not, I think the theme holder could generate one.

Caching by SVG path + theme would be enough to work, you'd have an instance of the Svg struct for each SVG + theme combo.
You'd use something like `std::tuple` or `std::pair` to hold a `std::string` and a `std::weak_ptr<Theme>`, and use these as the key for a map, and you'd define your own map like SvgCache as `static std::map<std::pair<std::string, std::weak_ptr<Theme>>, std::string> themedSvgCache;`


> I made some attempts at trying to make a custom Svg::load without modifying any of the Rack files, but ran into scoping problems or various flavours thereof (and that's when I gave up and just started recompiling Rack, heh). Something about that static svgCache in Svg.cpp is throwing me for a loop. It's highly likely that I just don't know what I'm doing; my c++ isn't very sophisticated.

svgCache isn't usable outside of Svg.cpp, you'd have to define your own. For caching tied to a theme like this, you'll also have to use a custom cache.

**Comment by landgrvi** [comment]
I see your point; that's certainly a more efficient use of the cache. 

It seems that applying the theme would have to change as well - rather than modifying the current svg, applying the theme would have to fetch the appropriate cached one.

**Comment by landgrvi** [comment]
I've got some stuff coded up based on the general principles discussed here that seems to be working; I'll see if I can make it less messy and turn it into a coherent PR. Might not be this week though!

**Comment by landgrvi** [comment]
I did do a pull request for this - hope it's usable!

**Comment by Paul-Dempsey** [comment]
I really appreciate the work you did. It might be a little bit before I get a chance to review and hopefully merge it. Might be a couple of weeks (in the new year), as I'm away from home with family for the holidays and transitioning my mother into hospice care.

**Comment by Paul-Dempsey** [comment]
Fixed by 4b17896
